src/material-moment-adapter/adapter/moment-date-adapter.ts

```diff
@@ -0,0 +1,178 @@
+@Injectable()
+export class MomentDateAdapter extends DateAdapter<moment.Moment> {
+  // Note: all of the methods that accept a `moment.Moment` input parameter immediately call
+  // `this.clone` on it. This is to ensure that we're working with a `moment.Moment` that has the
+  // correct locale setting while avoiding mutating the original object passed to us.
+
+  private _localeData: {
+    firstDayOfWeek: number,
+    longMonths: string[],
+    shortMonths: string[],
+    dates: string[],
+    longDaysOfWeek: string[],
+    shortDaysOfWeek: string[],
+    narrowDaysOfWeek: string[]
+  };
+
+  constructor(@Optional() @Inject(MAT_DATE_LOCALE) dateLocale: string) {
+    super();
+    this.setLocale(dateLocale || moment.locale());
+  }
+
+  setLocale(locale: any) {
+    super.setLocale(locale);
+
+    // Temporarily change the global locale to get the data we need, then change it back.
+    let globalLocale = moment.locale();
+    moment.locale(locale);
+    this._localeData = {
+      firstDayOfWeek: moment.localeData().firstDayOfWeek(),
+      longMonths: moment.months(),
+      shortMonths: moment.monthsShort(),
+      dates: range(31, (i) => this.createDate(2017, 0, i + 1).format('D')),
+      longDaysOfWeek: moment.weekdays(),
+      shortDaysOfWeek: moment.weekdaysShort(),
+      narrowDaysOfWeek: moment.weekdaysMin(),
+    };
+    moment.locale(globalLocale);
+  }
+
+  getYear(date: moment.Moment): number {
+    return this.clone(date).year();
+  }
+
+  getMonth(date: moment.Moment): number {
+    return this.clone(date).month();
+  }
+
+  getDate(date: moment.Moment): number {
+    return this.clone(date).date();
+  }
+
+  getDayOfWeek(date: moment.Moment): number {
+    return this.clone(date).day();
+  }
+
+  getMonthNames(style: 'long' | 'short' | 'narrow'): string[] {
+    // Moment.js doesn't support narrow month names, so we just use short if narrow is requested.
+    return style == 'long' ? this._localeData.longMonths : this._localeData.shortMonths;
+  }
+
+  getDateNames(): string[] {
+    return this._localeData.dates;
+  }
+
+  getDayOfWeekNames(style: 'long' | 'short' | 'narrow'): string[] {
+    if (style == 'long') {
+      return this._localeData.longDaysOfWeek;
+    }
+    if (style == 'short') {
+      return this._localeData.shortDaysOfWeek;
+    }
+    return this._localeData.narrowDaysOfWeek;
+  }
+
+  getYearName(date: moment.Moment): string {
+    return this.clone(date).format('YYYY');
+  }
+
+  getFirstDayOfWeek(): number {
+    return this._localeData.firstDayOfWeek;
+  }
+
+  getNumDaysInMonth(date: moment.Moment): number {
+    return this.clone(date).daysInMonth();
+  }
+
+  clone(date: moment.Moment): moment.Moment {
+    return date.clone().locale(this.locale);
+  }
+
+  createDate(year: number, month: number, date: number): moment.Moment {
+    // Check for invalid month and date (except upper bound on date which we have to check after
+    // creating the Date).
```

Comment:
Mention what moment would otherwise do if the month is out of range?

Reply:
Moment actually does the right thing, just want to be able to give better errors to the user